### PR TITLE
[Welcome] Post failed DMs to welcome channel

### LINKED ISSUE #515: [Welcome] Post DM message to welcome channel if DMs are disabled
If a new user joins a server with welcome DMs and welcome channel enabled but they have DMs disabled, we log that they have DMs disabled but don't attempt to notify them about this DM. We should additionally post this DM in the introductions channel **if the DM fails**, prefaced with something along the lines of:
```
Hey <at-mention>, we couldn't reach your DMs. Please review the following:
<insert embed here>
```

### PR DESCRIPTION
**Pull request by quachtridat**
## Summary

As described in #515, this PR adds a toggle which lets `Welcome` cog posts the embed that is supposed to be sent to new users coming to a server, but was failed (e.g., due to users' privacy settings blocking DMs).

### New command

- `[p]welcomeset channelconfig postfaileddm`: toggle the behavior. Short aliases are also available.

### New `async` method `sendToWelcomeChannel()`

This method posts a Discord message to a guild's welcome channel and raises exceptions accordingly on error.

### Additional logic in `sendWelcomeMessages`

On failed DMs, if the behavior is enabled, call `sendToWelcomeChannel()` with a small message to notify the respective new users about the DM welcome embed.

### New constants

- `KEY_WELCOME_CHANNEL_SETTINGS`
- `KEY_POST_FAILED_DM`

## Linked issues

This resolves #515.

## Notes

Might need some discussion about the command group `channelconfig` and `KEY_WELCOME_CHANNEL_SETTINGS`. They are probably redundant now, but it seems that *posting failed DMs* is only a *welcome channel* thing, and perhaps having a group for welcome-channel-related configs would make subcommands more organized (e.g., `[p]welcomeset greetings` subcommands may manage greetings and `[p]welcomeset greetings channel` may become `[p]welcomeset channelconfig greetingchannel` instead.

### CODE REVIEW

**Comment by Injabie3** [comment]
Re: command group, I think we might want to restructure the commands/groups under `[p]welcomeset`. Originally we had it for welcome DMs only, but with further additions, the command layout has become a bit messy.

Perhaps we could do something like:
- [p]welcomeset
    - dm
        - message
        - title
        - image
        - test
        - toggle
    - greetings
        - add
        - list
        - remove
        - toggle
    - log
        - channel
        - tag
            - add
            - get
            - list
            - remove
        - toggle


Open to inputs here.

**Comment by quachtridat** [comment]
> * [p]welcomeset
>   
>   * dm
>     
>     * message
>     * title
>     * image
>     * test
>     * toggle
>   * greetings
>     
>     * add
>     * list
>     * remove
>     * toggle
>   * log
>     
>     * channel
>     * tag
>       
>       * add
>       * get
>       * list
>       * remove
>     * toggle
>
> Open to inputs here.

Somehow I feel that `tag` and `log` would look better if they were at the same level. Also we need `[p]welcomeset greetings channel`. Everything else seems good to me.

Also if we opt in for `guildConfig[KEY_DM][KEY_MESSAGE]` (the nested `dict` way) we'll need to move stored things around as well.

**Comment by Injabie3** [comment]
Coming back to this, I think we should change the command hierarchy first in another PR, and then come back to this failed DM messaging thing after that's complete.

> Also we need `[p]welcomeset greetings` channel.

yep, thanks +1

> I feel that `tag` and `log` would look better if there were at the same level.

I'm indifferent, I put `tag` under `log` because tag uses the same "logging" channel.

**Comment by quachtridat** [comment]
> Coming back to this, I think we should change the command hierarchy first in another PR, and then come back to this failed DM messaging thing after that's complete.

I agree.

> I'm indifferent, I put `tag` under `log` because tag uses the same "logging" channel.

I would want to think that `log` and `tag` should be at the same level because tagging/description assigning does not sound like a sub-feature of logging.

**Comment by Injabie3** [comment]
Any updates on this one?

**Comment by Injabie3** [comment]
Or were you waiting for me to review it again?

**Comment by quachtridat** [comment]
> Any updates on this one?

Ah sorry, I haven't got a chance to make further changes. Again, this is what I am considering (https://github.com/SFUAnime/Ren/issues/533):

`[p]welcomeset`
* `dm`
    * `message`
    * `title`
    * `image`
    * `test`
    * `toggle`
* `greetings`
    * `add`
    * `channel`
    * `list`
    * `remove`
    * `toggle`
* `log`
    * `channel`
    * `toggle`
* `tag`/`desc`/`description`/`descriptions`
    * `add`
    * `get`
    * `list`
    * `remove`

Will see if I can do https://github.com/SFUAnime/Ren/issues/533 within the coming week.

File: cogs/welcome/welcome.py [hunk 1]
```diff
@@ -86,6 +86,48 @@ async def on_guild_channel_delete(self, removedChannel: discord.TextChannel):
...
+            The message sent
+
+        Raises
+        ------
+        discord.Forbidden
+            If the bot doesn't have permissions to send messages to the channel
+        discord.HTTPException
+            If the message couldn't be sent
+        discord.InvalidArgument
+            If keyword arguments are invalid
```
Comment (Injabie3) [inline cogs/welcome/welcome.py:114]: Please add a docstring comment for `ValueError`

**Review by Injabie3 — CHANGES_REQUESTED** [review]
As per sideband discussion, will wait for manual re-test confirmation.

**Comment by quachtridat** [comment]
> As per sideband discussion, will wait for manual re-test confirmation.

Re-tested @ https://discord.com/channels/800838291959447602/895950136688132137/985810413847461899.

**Review by Injabie3 — APPROVED** [review]

**Comment by Injabie3** [comment]
Will merge and cherry pick the squashed commit to the cog repo since welcome changes have diverged since the PR was opened.

**Pull request merged**